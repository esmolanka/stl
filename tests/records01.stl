
-- Empty records unify
#check {} <: {}

-- Any extra fields on the LHS unify with the polymoprphic tail on the RHS
#check {foo : Unit} <: {}
#check {foo+ : Unit} <: {}

-- Optional fields on the RHS unify with implicit 'Absent' fields on
-- the LHS
#check {} <: {foo : Unit}
-- 'Present' fields on the RHS _do_not_ unify with implicit 'Absent'
-- fields on the LHS
#check {} <: {foo+ : Unit}

-- Field order does not matter
#check {foo : Unit, bar : Void} <: {bar : Void, foo : Unit}
#check {foo+ : Unit, bar+ : Unit} <: {bar+ : Unit, foo+ : Unit}

-- For extra fields, too
#check {foo+ : Unit, baz : Unit, bar+ : Unit} <: {bar+ : Unit, foo+ : Unit}
#check {foo+ : Unit, bar+ : Unit} <: {bar+ : Unit, baz : Unit, foo+ : Unit}

-- Matching fields' types must unify
#check {foo : Unit, bar : Void} <: {bar : Unit, foo : Unit}
#check {foo+ : Unit, bar+ : Void} <: {bar+ : Unit, foo+ : Unit}
#check {foo+ : Unit, bar+ : {baz : Unit}} <: {bar+ : {quux : Unit}, foo+ : Unit}

