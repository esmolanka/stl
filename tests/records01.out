tests/records01.stl:3:1-3:16:
  Subtype:
    ∃ (ρ : Row). {ρ}
  Supertype:
    ∃ (ρ : Row). {ρ}
  Result:
    OK
tests/records01.stl:6:1-6:26:
  Subtype:
    ∃ (ρ : Row). (∃ (ω : #). {foo^ω : Unit | ρ})
  Supertype:
    ∃ (ρ : Row). {ρ}
  Result:
    OK
tests/records01.stl:7:1-7:27:
  Subtype:
    ∃ (ρ : Row). {foo : Unit | ρ}
  Supertype:
    ∃ (ρ : Row). {ρ}
  Result:
    OK
tests/records01.stl:11:1-11:26:
  Subtype:
    ∃ (ρ : Row). {ρ}
  Supertype:
    ∃ (ρ : Row). (∃ (ω : #). {foo^ω : Unit | ρ})
  Result:
    OK
tests/records01.stl:14:1-14:27:
  Subtype:
    ∃ (ρ : Row). {ρ}
  Supertype:
    ∃ (ρ : Row). {foo : Unit | ρ}
  Result:
    FAIL
    Cannot unify types.
        ▫︎
      Is not subsumed by:
        ▪︎
tests/records01.stl:17:1-17:60:
  Subtype:
    ∃ (ρ : Row). (∃ (ω : #). (∃ (ω : #). {foo^ω : Unit, bar^ω/1 : Void | ρ}))
  Supertype:
    ∃ (ρ : Row). (∃ (ω : #). (∃ (ω : #). {bar^ω : Void, foo^ω/1 : Unit | ρ}))
  Result:
    OK
tests/records01.stl:18:1-18:64:
  Subtype:
    ∃ (ρ : Row). {foo : Unit, bar : Unit | ρ}
  Supertype:
    ∃ (ρ : Row). {bar : Unit, foo : Unit | ρ}
  Result:
    OK
tests/records01.stl:21:1-21:76:
  Subtype:
    ∃ (ρ : Row). (∃ (ω : #). {foo : Unit, baz^ω : Unit, bar : Unit | ρ})
  Supertype:
    ∃ (ρ : Row). {bar : Unit, foo : Unit | ρ}
  Result:
    OK
tests/records01.stl:22:1-22:76:
  Subtype:
    ∃ (ρ : Row). {foo : Unit, bar : Unit | ρ}
  Supertype:
    ∃ (ρ : Row). (∃ (ω : #). {bar : Unit, baz^ω : Unit, foo : Unit | ρ})
  Result:
    OK
tests/records01.stl:25:1-25:60:
  Subtype:
    ∃ (ρ : Row). (∃ (ω : #). (∃ (ω : #). {foo^ω : Unit, bar^ω/1 : Void | ρ}))
  Supertype:
    ∃ (ρ : Row). (∃ (ω : #). (∃ (ω : #). {bar^ω : Unit, foo^ω/1 : Unit | ρ}))
  Result:
    FAIL
    Cannot unify types.
        Void
      Is not subsumed by:
        Unit
tests/records01.stl:26:1-26:64:
  Subtype:
    ∃ (ρ : Row). {foo : Unit, bar : Void | ρ}
  Supertype:
    ∃ (ρ : Row). {bar : Unit, foo : Unit | ρ}
  Result:
    FAIL
    Cannot unify types.
        Void
      Is not subsumed by:
        Unit
tests/records01.stl:27:1-27:81:
  Subtype:
    ∃ (ρ : Row). {foo : Unit, bar : ∃ (ρ : Row). (∃ (ω : #). {baz^ω : Unit | ρ}) | ρ}
  Supertype:
    ∃ (ρ : Row). {bar : ∃ (ρ : Row). (∃ (ω : #). {quux^ω : Unit | ρ}), foo : Unit | ρ}
  Result:
    OK